### PR TITLE
Build Prototype Monkey Island AppImage

### PR DESCRIPTION
**Pull request by mssalvatore**
# What does this PR do? 

Adds a deployment script that builds Monkey Island as a prototype AppImage.
The changes to the python code have already been reviewed in the following PRs:
- #963
- #971
- #994
- #997

These changes were rebased on the develop branch.

The only changes that have not been reviewed are the ones in the `deployment_scripts` directory.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing? 
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by building and running an AppImage locally.
    > Tested by running blackbox tests against AppImage
* [ ] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

**Comment by VakarisZ** [comment]
Are there any pre-requisites not mentioned? Or can I literraly just run `./build_appimage.sh` from any linux and it will work? Maybe add a readme, even if it's just one line of text saying "run ./build_appimage.sh"?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:14]: Do we want to remove all these dirs once appimage is built?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:42]: What if distro uses wget instead of curl? That's the things we can avoid specifying which distros can be used to build app-image in docs

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:194]: Pretty much looks like config in deployment script. You use the url because script hasn't downloaded the repo yet? My questions then: 1. why don't use a file from our repo? 2. If user is using this script without the whole repo, how did he get this script? Pulled only a specific folder? If we had usage docs, would be easier to understand

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:5]: Is there a specific reason we want `/usr/src`?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:54]: maybe add `sudo apt update` as well?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:73]: Maybe make this link a const as well, since it's likelly to get updated

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:30]: why have`()` if it can't contain parameters?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:49]: Export to a const as well, because the version of nodejs might change?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:95]: will get re-targeted in a different PR? Make a const?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:108]: Why not move the whole dir? Saving disc space?

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:121]: `line_with_pyinstaller=$(cat $requirements_island | grep -i -n pyinstaller)`

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:120]: I think this applies to virtualenv as well

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:175]: This means that monkeys can only be executed on the island or what?

Comment (VakarisZ) [reply deployment_scripts/appimage/build_appimage.sh:175]: Ahh, this is the explanation to why you download the binaries after building the AppDir

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:190]: It's something that would look good in docs and save the user some trial and error

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:220]: why do we need this as a separate command?

File: deployment_scripts/appimage/monkey_island.desktop
Comment (VakarisZ) [inline deployment_scripts/appimage/monkey_island.desktop:5]: Fill him in?

File: deployment_scripts/appimage/run_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/run_appimage.sh:1]: Can you explain when this file is used and when AppRun is used?

**Review by VakarisZ — CHANGES_REQUESTED** [review]
Phenomenal quality of bash scripts. Some duplication with linux deployment scripts and maybe that can be resolved in the future. 
Left some questions, some arise due to the lack of knowledge about how appimage works, others are general suggestions.

File: CHANGELOG.md
Comment (mssalvatore) [reply CHANGELOG.md:10]: This is still in the testing phase. Regardless, I'd prefer to put the change log entries in chronological order.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:14]: Maybe, but I'm envisioning this being run in a clean VM or LXC and then being torn down. I'll add a README with this information.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:194]: I'll add a readme. I pretty much copied and reworked the linux deployment script.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:5]: It's an AppImage convention.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:59]: wget is in ubuntu by default. Curl didn't seem to be. I'll add details about the requirements of the build environment in a readme.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:54]: The VM should be upgraded before building. I'll add details in a readme.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:30]: It's the Bourne syntax instead of the POSIX syntax, i.e. `my_func()` vs `function my_func`. I don't care much. I can change them all if you want.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:95]: I'll just switch it to develop for now.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:175]: Yep. There are better solutions to this problem, but since we're not using AppImage Builder in the long run, they're not worth investing in.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:190]: I'll add notes to the readme. The message is still useful if you forget.

File: deployment_scripts/appimage/monkey_island.desktop
Comment (mssalvatore) [reply deployment_scripts/appimage/monkey_island.desktop:5]: AppImage Builder doesn't actually use this file, but different packaging methods do. I'll fill this in in round 2 or remove this wile.

File: deployment_scripts/appimage/run_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/run_appimage.sh:1]: AppRun is part of the AppImage spec. It's the entry point. AppImage Builder generates its own AppRun from the `exec` and `exec_args` fields in `monkey_island_builder.yml`. ATM, AppRun isn't used, so I'll just remove it.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:14]: Oh, also, those are subdirectories in $APPDIR, which you generally want to remove between runs, but for development and debugging you don't.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:108]: There are build requirements (like monkey_island_builder.yaml) and build artifacts (like an appimage-builder cache directory) that don't need to be in the final appimage. Also, these files get copied to specific places. It's not ideal, but can be reworked when we chose a different build tool.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:14]: See the new README for info on cleaning up.

File: deployment_scripts/appimage/build_appimage.sh
Comment (acepace) [reply deployment_scripts/appimage/build_appimage.sh:30]: If you're wondering, can probably defer to shellcheck

Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:30]: Shellcheck doesn't complain about it. Mostly it complains about my lack of double quotes around variables in or containing paths.

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [reply deployment_scripts/appimage/build_appimage.sh:56]: Even if it's docker or something, how do you know that the repositories will be updated? I'd rather we have `sudo apt update` in scripts than specify in the readme that "you need to run sudo apt update".

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [reply deployment_scripts/appimage/build_appimage.sh:30]: No need to change

File: deployment_scripts/appimage/build_appimage.sh
Comment (VakarisZ) [inline deployment_scripts/appimage/build_appimage.sh:48]: If you move links like these to the top of the file, it will be easier to fix once they get broken, since you won't have to scroll the whole file. Not worth holding up the PR though

File: deployment_scripts/appimage/README.md
Comment (VakarisZ) [inline deployment_scripts/appimage/README.md:11]: Any reason why this shouldn't be a part of build script?

File: deployment_scripts/appimage/README.md
Comment (VakarisZ) [inline deployment_scripts/appimage/README.md:18]: Woulnd't it be easier in terms of integration testing, to NOT require a new machine each time?

**Review by VakarisZ — APPROVED** [review]
Small comments and questions, but can be solved later/ignored

File: deployment_scripts/appimage/README.md
Comment (mssalvatore) [reply deployment_scripts/appimage/README.md:11]: Arguably yes, but in its current form, no. I'll add it to the build script. We can revisit it later.

File: deployment_scripts/appimage/README.md
Comment (mssalvatore) [reply deployment_scripts/appimage/README.md:18]: LXCs are very quick to spin up. So are VMs in the right environment. Regardless, the README contains instructions on how to clean up after the script runs so that you don't need to use a fresh environment each time.

File: deployment_scripts/appimage/build_appimage.sh
Comment (mssalvatore) [reply deployment_scripts/appimage/build_appimage.sh:48]: I like having it localized, as that function is the only place it's used, but I see your point. I'll leave it for now, as this may get reworked anyway in version 2.

**Pull request merged**
